### ISSUE TITLE
Extension for scipy.spatial.distance.cdist

### ISSUE THREAD

**Issue by YubinXie** [scipy/scipy#10181]
scipy.spatial.distance.cdist is a great function that provides fast calculation of pairwise distance between two lists of points.

I want to get the pairwise min distance between many lists of points. By using cdist in two for loop function is very slow``` [[ np.min(cdist(x,y)) for y in list ]for x in list]```

I was wondering if there are any extension function like cdist not just to get distance between two lists of points, but to get get pairwise distance between many lists of points.

Thanks!

**Comment by tylerjereddy** [comment]
A binary space partitioning tree might be helpful.

[`scipy.spatial.cKDTree.query`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.query.html#scipy.spatial.cKDTree.query) or similar once you've initialized the k-d tree data structure. You'd still need to loop, I think, but there's built-in low-level code for finding the pairwise nearest neighbors.

That may also have the advantage of not scaling the space complexity as much, since those `cdist` results can get pretty big for the case where you only want the min dist.

**Comment by YubinXie** [comment]
I tried the cKDTree as followed (let me know if I did something wrong here):

![04CDE154-785B-4FF0-9ED5-3314CEACD03E](https://user-images.githubusercontent.com/16257776/57807467-3baef200-772f-11e9-9c5a-7c9f275d8f34.png)

Maybe because my list of points are too large (2000 lists), it seems building a tree for each loop is not the best strategy.

**Comment by peterbell10** [comment]
You should build each tree only once and then query it multiple times:
```python
trees = [cKDTree(x) for x in ellipse_list]
[[ x.query(y) for y in ellipse_list] for x in trees]
```

**Comment by YubinXie** [comment]
Thanks for your suggestion! I tried again and it is slightly faster. I feel probably it is better for me to constrain the possible pairs instead of use the full pair-wise. 

![549E16C8-9963-4F7C-848C-CEC30373D308](https://user-images.githubusercontent.com/16257776/57811655-d102b400-7738-11e9-8625-aeae48e34fbc.png)

**Comment by tylerjereddy** [comment]
A few more thoughts:

- Shouldn't you only need `x-1` trees? If there are 3 lists of points `x` in `ellipse_list`, you can get the requisite min dist comparisons with only two trees and queries on the last list of points since you shouldn't have to query List3->1 or List3->2 after doing list1->3, list2->3.
- what about heuristics like the `distance_upper_bound` argument to prune the tree searches? if you only care about min dist, then just pick any random pair of points from a set being compared as the cap, since that is either the minimum or some real distance cap above it; if the data sets are highly structured / orderd, shuffling might also help, but that requires more testing
- you might also be able to get some intelligent caps on min distance because of things like knowing the distance between L1 and L2, and L3 and L2, so that min dist between L3 and L1 has a cap of the sum of the other minimums to be used for `distance_upper_bound`; these relationships might even compound / reduce in interesting ways with more lists and relationships between them
- could also consider `n_jobs` argument to use multiple cores or something like dask to delay the evaluation and perform on multiple cores after getting reasonable guesses for `distance_upper_bound`, maybe

**Comment by YubinXie** [comment]
Thank you for your suggestion. I think I will limit the combination number first based on proximity distance.

One thing to mention is that, for the n_jobs, I tried it, and the running time increased from 3 min to 1h30 min for unknown reason when I set n_jobs=-1.

**Comment by rgommers** [comment]
> One thing to mention is that, for the n_jobs, I tried it, and the running time increased from 3 min to 1h30 min for unknown reason when I set n_jobs=-1.

this sounds like a bug. if you could open a new issue with a reproducible example, that would be useful
